File: 1200-minimum-absolute-difference/1200-minimum-absolute-difference.js
/**
 * @param {number[]} arr
 * @return {number[][]}
 */
// var minimumAbsDifference = function(arr) {
// //     sort arr
// //     get the adjecnt two element difference to find the minimum
//     let sortedArr= arr.sort((a,b)=>a-b)
//     let minDiff=Infinity
//     let res=[]
//     for(let i=0;i<arr.length-1;i++){
//         let diff=Math.abs(sortedArr[i+1]-sortedArr[i])
//         if(diff<minDiff){
//             minDiff=diff
           
//         }
       
//     }
//     for(let i=0;i<sortedArr.length;i++){
//         for(let j=1;j<sortedArr.length;j++){
//             if(sortedArr[j]-sortedArr[i]===minDiff) res.push([sortedArr[i],sortedArr[j]])
//         }
//     }
    
//     return res
// };
var minimumAbsDifference = function(arr) {
    // Sort the array
    arr.sort((a, b) => a - b);

    let minDiff = Infinity;
    let res = [];

    // Find the minimum difference
    for (let i = 0; i < arr.length - 1; i++) {
        minDiff = Math.min(minDiff, arr[i + 1] - arr[i]);
    }

    // Find all pairs with the minimum difference
    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i + 1] - arr[i] === minDiff) {
            res.push([arr[i], arr[i + 1]]);
        }
    }

    return res;
};

File: 1200-minimum-absolute-difference/1200-minimum-absolute-difference.py
class Solution:
    def minimumAbsDifference(self, arr: List[int]) -> List[List[int]]:
#         brute force O(n^2)
        # min_diff=math.inf
        # arr.sort()
        # res=[]
        # for i in range(len(arr)-1):
        #     for j in range(i+1,len(arr)):
        #         min_diff=min(min_diff,abs(arr[i]-arr[j]))
        # for i in range(len(arr)-1):
        #     for j in range(i+1,len(arr)):
        #         if abs(arr[i]-arr[j])==min_diff:
        #             res.append([arr[i],arr[j]])
        # return res
        
#         optimize
#         because the arr will be sorted=>the diff between to nearest two num will be smaller than non-neighbor ones
#         no need to check all possible pairs
        arr.sort()
        min_diff=math.inf
        for i in range(len(arr)-1):
            min_diff=min(min_diff,abs(arr[i]-arr[i+1]))
            
        res=[]
        for i in range(len(arr)-1):
            if abs(arr[i]-arr[i+1])==min_diff:
                res.append([arr[i],arr[i+1]])
        return res
        